interface/modules/custom_modules/oe-module-comlink-telehealth/src/Controller/TeleconferenceRoomController.php

```diff
@@ -0,0 +1,781 @@
+        try {
+            if (!empty($session)) {
+                $result['session']['id'] = $session['id'];
+                // provider has started the session, let's verify the last update time
+                if (!empty($session['provider_last_update'])) {
+                    $dateTime = \DateTime::createFromFormat("Y-m-d H:i:s", $session['provider_last_update']);
+                    $currentDateTime = new \DateTime();
+                    $this->logger->debug("checking time ", ['provider_last_update' => $currentDateTime->format("Y-m-d H:i:s"), 'now' => $currentDateTime->format("Y-m-d H:i:s")]);
+                    if ($currentDateTime < $dateTime->add(new \DateInterval("PT15S"))) {
+                        $result['session']['providerReady'] = true;
+                    }
+                    $userRepo = new UserService();
+                    $user = $userRepo->getUser($session['user_id']);
+                    if (empty($user)) {
+                        throw new InvalidArgumentException("Failed to get user with session id " . $session['id']);
+                    }
+
+                    // in the event that we never change the login credentials we grab it from the session here...
+                    $telehealthCredentials = $this->telehealthUserRepo->getUser($user['uuid']);
+                    if (empty($telehealthCredentials)) {
+                        throw new InvalidArgumentException("Failed to get telehealth credentials with username " . $user['username']);
+                    }
+                    $result['session']['calleeUuid'] = $telehealthCredentials->getUsername();
+                }
+            }
+            echo json_encode($result);
+        } catch (Exception $exception) {
+            (new SystemLogger())->errorLogCaller($exception->getMessage(), ['trace' => $exception->getTraceAsString(),
+                'queryVars' => $queryVars]);
+            http_response_code(500);
+            echo json_encode(['error' => 'server error occurred.  Check server logs for details']);
+        }
+    }
+
+    public function checkRegistrationAction($queryVars)
+    {
+        try {
+            if ($this->isPatient) {
+                $patient = $this->getPatientForPid($queryVars['pid']);
+                $settings = $this->getOrCreateTelehealthPatient($patient);
+            } else {
+                $userRepo = new UserService();
+                $user = $userRepo->getUserByUsername($queryVars['authUser']);
+                $settings = $this->getOrCreateTelehealthProvider($user);
+            }
+            $jsonSettings = json_encode([
+                'username' => $settings->getUsername()
+                , 'isPatient' => $this->isPatient
+                , 'dbRecordId' => $settings->getId()
+            ]);
+            $this->logger->debug("check registration finished ", ['settings' => $jsonSettings]);
+            echo $jsonSettings;
```

Comment:
when echoing json_encoded stuff in html context, recommend wrapping it with text()
(looks like several places where this is missing)

Reply:
Are you suggesting I do this? text(json_encode(..))? Won't this break the json_encode?

Reply:
we have done this in many places (since it is vulnerable to html xss attacks if in html context) and no breakage yet :) (if the script is set to json only then i don't think needed and then prob more risk for breakage (although not sure)

Reply:
Yea, all of these places I'm echoing JSON its because its being used by an AJAX call.  Is the proper procedure here to do a jsText and jsAttr on the frontend where these are being consumed?

Reply:
no, no need to deal with in front end script if it ends up in js context. the main issue is when this script echoes this php string if there is a way to directly call it via a hard weblink (most ajax scripts there is a way) (rec putting a bad first name with `alert(document.cookie)` and then calling it directly via a hard weblink and seeing if it throws the alert)

Reply:
sorry, meant to write `<script>alert(document.cookie)</script>` for the bad first name

Reply:
@bradymiller so I'm actually running into problems using this method with the following code segment on L211-L218
```
 $result = [
                // waiting room has already been escaped via twig rendering
                'waitingRoom' => $waitingRoom
                , 'callerSettings' => $settings
                // conference room has already been escaped via twig rendering
                , 'conferenceRoom' => $conferenceRoom
            ];
            echo text(json_encode($result));
```
It ends up mangling the html which can no longer be injected into the DOM via javascript.  $waitingRoom and $conferenceRoom are already being escaped by going through the $twig->render() functionality.  I'm thinking of just escaping each of the individual array values in the 'callerSettings' property to accomplish what I need to sanitize the output.  Do you see problems with this approach?

Reply:
To clarify, is there html entities in these variables?

What is the direct link that will show the output for the getTeleHealthLaunchDataAction() function (ie. what ajax link the main script uses to make the call). Following this link directly from the browser will demonstrate how most these vulnerabilities happen. Meaning, they usually call the ajax script directly through the web browser (ie. so even if it escaped downstream in the client twig script, still vulnerable). And for some reason double escaping does not happen if do this (ie. escape the ajax output and the client output that is calling the ajax output).

Reply:
I just answered the first question by looking at the code. Looks like sending back twig rendered output with html entities in the array, so yep can't escape that :)
(when have html entities embedded in strings, then basically need to do the escaping upstream where the string is built like you are doing)

Reply:
And in this case, agree the textArray is a good way to then deal with the non-html entity string that needs escaping

Reply:
good marking as resolved.